packages/xrpl/src/sugar/submit.ts

```diff
@@ -119,29 +120,42 @@ async function submitRequest(
 async function waitForFinalTransactionOutcome(
   client: Client,
   txHash: string,
+  lastLedger: number,
 ): Promise<TxResponse> {
   await sleep(LEDGER_CLOSE_TIME)
 
-  const txResponse = await client.request({
-    command: 'tx',
-    transaction: txHash,
-  })
+  const latestLedger = await client.getLedgerIndex()
+
+  const txResponse = await client
+    .request({
+      command: 'tx',
+      transaction: txHash,
+    })
+    .catch(async (error) => {
+      // this type of error does not have its interface, type-assert and extract the value we need.
+      // eslint-disable-next-line @typescript-eslint/consistent-type-assertions,@typescript-eslint/no-unsafe-member-access -- L131
+      const message = error.data.error as string
+      if (message === 'txnNotFound') {
+        if (lastLedger > (await client.getLedgerIndex())) {
+          throw new XrplError(
+            `The latest ledger sequence ${latestLedger} is greater than the transaction's LastLedgerSequence (${lastLedger}).`,
+          )
+        }
+        return waitForFinalTransactionOutcome(client, txHash, lastLedger)
+      }
+      throw new Error(message)
+    })
+
   if (txResponse.result.validated) {
     return txResponse
   }
 
-  const txLastLedger = txResponse.result.LastLedgerSequence
-  if (txLastLedger == null) {
-    throw new XrplError('LastLedgerSequence cannot be null')
-  }
-  const latestLedger = await client.getLedgerIndex()
-
-  if (txLastLedger > latestLedger) {
-    return waitForFinalTransactionOutcome(client, txHash)
+  if (lastLedger > latestLedger) {
```

Comment:
Same

Reply:
not this, here we are making the call again, only if the last ledger seq is greater

Reply:
Ah, you're right, because this one is recursing. Good catch.